src/main/kotlin/org/jboss/tools/intellij/kubernetes/model/resource/AbstractResourcesProvider.kt

```diff
@@ -10,45 +10,70 @@
 
+import com.intellij.openapi.diagnostic.logger
 import io.fabric8.kubernetes.api.model.HasMetadata
-import org.jboss.tools.intellij.kubernetes.model.util.areEqual
+import org.jboss.tools.intellij.kubernetes.model.util.sameUid
 
 abstract class AbstractResourcesProvider<R : HasMetadata> : IResourcesProvider<R> {
 
-    protected val allResources = mutableSetOf<R>()
+    protected val allResources = mutableListOf<R>()
 
     override fun invalidate() {
+        logger<AbstractResourcesProvider<*>>().debug("Invalidating all $kind resources.")
         synchronized(allResources) {
             allResources.clear()
         }
     }
 
-    override fun invalidate(resource: HasMetadata) {
-        synchronized(allResources) {
-            allResources.remove(resource)
-        }
-    }
-
     override fun add(resource: HasMetadata): Boolean {
-        if (!kind.clazz.isAssignableFrom(resource::class.java)) {
+        if (!isOfCorrectKind(resource)) {
             return false
         }
+        logger<AbstractResourcesProvider<*>>().debug("Adding resource ${resource.metadata.name}.")
         // don't add resource if different instance of same resource is already contained
         synchronized(allResources) {
-            return allResources.find { areEqual(it, resource) } == null
-                    && allResources.add(resource as R)
+            val existing = allResources.find { resource.sameUid(it) }
+            return when (existing) {
+                resource -> false
+                null -> allResources.add(resource as R)
+                else -> replaceBy(existing, resource)
+            }
         }
     }
 
     override fun remove(resource: HasMetadata): Boolean {
-        val resourceClass = resource::class.java
-        if (!kind.clazz.isAssignableFrom(resourceClass)) {
+        if (!isOfCorrectKind(resource)) {
             return false
         }
         // do not remove by instance equality bcs instance to be removed can be different
         // ex. when removal is triggered by resource watch
+        logger<AbstractResourcesProvider<*>>().debug("Removing resource ${resource.metadata.name}.")
+        synchronized(allResources) {
+            return allResources.removeIf { resource.sameUid(it) }
+        }
+    }
+
+    override fun replace(resource: HasMetadata): Boolean {
+        if (!isOfCorrectKind(resource)) {
+            return false
+        }
+        logger<AbstractResourcesProvider<*>>().debug("Replacing resource ${resource.metadata.name}.")
         synchronized(allResources) {
-            return allResources.removeIf { areEqual(it, resource) }
+            val toReplace = allResources.find { resource.sameUid(it) } ?: return false
+            return replaceBy(toReplace, resource)
+        }
+    }
+
+    private fun replaceBy(toReplace: R, replaceBy: HasMetadata): Boolean {
+        val indexOf = allResources.indexOf(toReplace)
+        if (indexOf < 0) {
```

Comment:
I guess you can just write `allResources[allResources.indexOf(toReplace)] = replaceBy as R`. This is a private method only called by the `replace` above and before you call this you already check if the element exists in the list otherwise return false. BTW not so importat

Reply:
Wouldn't trying to set an element at a negative index throw an exception? In Java it would, not sure about kotlin yet.

Reply:
Trying to set an element at a negative index throws:
```
java.lang.IndexOutOfBoundsException: Index -1 out of bounds for length 3
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Preconditions.java:64)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Preconditions.java:70)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Preconditions.java:248)
```
Missing something?

Reply:
gotcha, you think that we could rely on the result of the find that we do right above. +1

Reply:
What i meant was. By using `allResources.find { resource.sameUid(it) } ?: return false` you already verify if the element exists inside the collection (~index >=0) otherwise you never call the `replaceBy` function.

```
override fun replace(resource: HasMetadata): Boolean {
        if (!isOfCorrectKind(resource)) {
            return false
        }
        logger<AbstractResourcesProvider<*>>().debug("Replacing resource ${resource.metadata.name}.")
        synchronized(allResources) {
           ----> val toReplace = allResources.find { resource.sameUid(it) } ?: return false
            return replaceBy(toReplace, resource)
        }
    }
```

Inside the replaceBy you check again if index is greater than -1 but you already did this check above. It will never happen you will have a -1 as if the element doesn't exist this method will never be called, right? 
```
private fun replaceBy(toReplace: R, replaceBy: HasMetadata): Boolean {
        val indexOf = allResources.indexOf(toReplace) 
        if (indexOf < 0) { --> check if the element exists but you already made this check above
```

other thing is that when you use the indexOf you don't check the UID and i guess the check could return a false positive(???) bc maybe the object you save is different from the one you are checking.
I think it may happen 
```
Pod
uid: aaaa
normal config
``` 

then k8s does something

```
Pod
uid: aaaa
normal config
stuff added by k8s for some reason
```

I guess at this point the indexOf will fail but i'm just guessing.

Reply:
I was writing while you already answered, sorry. Yes you got what i meant :laughing:. Give also a look at the indexOf, read above. Not sure how it works

Reply:
thinking twice I'd still prefer the current solution bcs it's more robust. The method would not rely on any external find that should happen so that it would not throw. At the same it, there's no perf hit. On the other hand I'd prefer not to have to find first and then look up again via index. I'd love to get the index when finding. Didnt' see how I'd achieve this.

Reply:
Yeah, thinking about the indexOf it always works fine in the current use case as the object checked is the one found by 'find'. Resolving it then.